### PR TITLE
Migrate state_dict bc test to OptimizerInfo, increase coverage

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116509
* __->__ #116500

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -448,6 +448,44 @@ def fwd_bwd(optim, w, b, i):
                 optimizer_c.state_dict()["param_groups"][-1]
             )
 
+    @optims(optim_db, dtypes=[torch.float32])
+    def test_bc_for_state_dict(self, device, dtype, optim_info):
```
Comment (albanD) [inline test/test_optim.py:452]: Should we run part of this as part of https://github.com/pytorch/pytorch/blob/80c07df659362a95da7cd4f3ec367abfdace38c4/.ci/pytorch/test.sh#L835 ?

We can dump state dicts for all models with the old version and load them with the new one!

Bonus point if you make it such that @mikaylagawarecki can easily add the same thing for all Module via ModuleInfo later :)

Comment (janeyx99) [reply test/test_optim.py:452]: This feels like a larger project and I agree we most likely want e2e tests for load_state_dict somewhere. I'm not sure this exact `test_forward_backward_compatibility` makes the most sense with regards to optimizers though...

Comment (janeyx99) [reply test/test_optim.py]: Okay, after a discussion offline, this ask has been scoped into a good BE task that I've now encapsulated in https://github.com/pytorch/pytorch/issues/116502.

I agree it isn't hard to do but is different enough from this PR.

**Review by albanD — APPROVED** [review]
Sound good!

**Pull request closed without merging**